### PR TITLE
Implement transitive pinning in new dependency graph resolver

### PR DESCRIPTION
**Pull request by jeffkl**
## Description

This adds the transitive pinning feature to the new dependency resolution algorithm.

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [ ] ~~Added tests~~ Existing tests cover scenario
- [ ] ~~Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.~~ Not needed

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs [hunk 1]
```diff
@@ -88,6 +91,8 @@ public DependencyGraphResolver(RestoreCollectorLogger logger, RestoreRequest res
                     hasInstallBeenCalledAlready = true;
                 }
 
+                TargetFrameworkInformation? projectTargetFramework = _request.Project.TargetFrameworks.FirstOrDefault(i => NuGetFramework.Comparer.Equals(i.FrameworkName, pair.Framework));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs:94]: There's already extension methods for this: https://github.com/NuGet/NuGet.Client/blob/f06d028edc7851dff7014fdca5e360667d72e8f7/src/NuGet.Core/NuGet.ProjectModel/PackageSpecExtensions.cs#L15-L25.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs [hunk 2]
```diff
@@ -112,8 +117,7 @@ public DependencyGraphResolver(RestoreCollectorLogger logger, RestoreRequest res
                     // This is guaranteed to be computed before any graph with a RID, so we can assume this will return a value.
 
                     // PCL Projects with Supports have a runtime graph but no matching framework.
-                    var runtimeGraphPath = _request.Project.TargetFrameworks.
-                            FirstOrDefault(e => NuGetFramework.Comparer.Equals(e.FrameworkName, tfmNonRidGraph.Framework))?.RuntimeIdentifierGraphPath;
+                    var runtimeGraphPath = projectTargetFramework?.RuntimeIdentifierGraphPath;
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs:120]: projectTargetFramework probably sohuldn't be null. if it's null, we'd likely have other issues.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs [hunk 3]
```diff
@@ -750,10 +807,25 @@ public DependencyGraphResolver(RestoreCollectorLogger logger, RestoreRequest res
...
+                            {
+                                newGraphNode.OuterNode = currentGraphNode;
+                                currentGraphNode.InnerNodes.Add(newGraphNode);
+                            }
+                            if (isCentralPackageTransitivePinningEnabled && !downgrades.ContainsKey(chosenItemRangeIndex) && !RemoteDependencyWalker.IsGreaterThanOrEqualTo(chosenItem.LibraryDependency.LibraryRange.VersionRange, dep.LibraryRange.VersionRange))
+                            {
+                                downgrades.Add(chosenItem.LibraryRangeIndex, (currentGraphNode, dep));
+                            }
+
+                            if (newGraphNode.Item.Key.Type != LibraryType.Project && newGraphNode.Item.Key.Type != LibraryType.ExternalProject && newGraphNode.Item.Key.Type != LibraryType.Unresolved && !versionConflicts.ContainsKey(chosenItemRangeIndex) && dep.SuppressParent != LibraryIncludeFlags.All && dep.LibraryRange.VersionRange != null && !dep.LibraryRange.VersionRange!.Satisfies(newGraphNode.Item.Key.Version) && !downgrades.ContainsKey(chosenItemRangeIndex))
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs:828]: I have a really really tough time processing this. 

Comments would be helpful for conditions that include this many and statements.

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs [hunk 4]
```diff
@@ -27,17 +27,31 @@ public class PackageSpecReferenceDependencyProvider : IDependencyProvider
...
 
+        private readonly bool _useLegacyDependencyGraphResolution = false;
+
         public PackageSpecReferenceDependencyProvider(
             IEnumerable<ExternalProjectReference> externalProjects,
             ILogger logger) :
             this(externalProjects,
-                environmentVariableReader: EnvironmentVariableWrapper.Instance)
+                environmentVariableReader: EnvironmentVariableWrapper.Instance,
+                useLegacyDependencyGraphResolution: false)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:37]: The package spec has the data about the new algorithm, you shouldn't need to pass it here.

**Review by nkolev92 — APPROVED** [review]
I know this has undergone testing and I trust our automated testing. I don't really have major feedback on the implementaiton, as I can't fully follow this. 

Left some non blocking comments that we can address in a follow up.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs (hunk 1 above)
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs:94]: I'll do this in the refactor and tracking it in https://github.com/NuGet/Client.Engineering/issues/2959

**Pull request merged**